h2/src/main/org/h2/util/DateTimeFunctions.java

```diff
@@ -374,6 +381,166 @@ public static Value extract(String part, Value value) {
+        // result to nanoseconds.
+        switch (timeUnit) {
+
+        case MICROSECOND:
+
+            long nanoInMicroSecond = 1_000l;
+            long microseconds = timeNanosRetrieved / nanoInMicroSecond;
+            timeNanos = microseconds * nanoInMicroSecond;
+            break;
+
+        case MILLISECOND:
+
+            long nanoInMilliSecond = 1_000_000l;
+            long milliseconds = timeNanosRetrieved / nanoInMilliSecond;
+            timeNanos = milliseconds * nanoInMilliSecond;
+            break;
+
+        case SECOND:
+
+            long nanoInSecond = 1_000_000_000l;
+            long seconds = timeNanosRetrieved / nanoInSecond;
+            timeNanos = seconds * nanoInSecond;
+            break;
+
+        case MINUTE:
+
+            long nanoInMinute = 60_000_000_000l;
+            long minutes = timeNanosRetrieved / nanoInMinute;
+            timeNanos = minutes * nanoInMinute;
+            break;
+
+        case HOUR:
+
+            long nanoInHour = 3_600_000_000_000l;
+            long hours = timeNanosRetrieved / nanoInHour;
+            timeNanos = hours * nanoInHour;
+            break;
+
+        case DAY_OF_MONTH:
+
+            timeNanos = 0l;
+            break;
+
+        case WEEK:
+
+            long absoluteDay = DateTimeUtils.absoluteDayFromDateValue(dateValue);
+            int dayOfWeek = DateTimeUtils.getDayOfWeekFromAbsolute(absoluteDay, 1);
+            if (dayOfWeek != 1) {
+                dateValue = DateTimeUtils.dateValueFromAbsoluteDay(absoluteDay - dayOfWeek + 1);
+            }
+            timeNanos = 0l;
+            break;
+
+        case MONTH:
+
+            long year = DateTimeUtils.yearFromDateValue(dateValue);
+            int month = DateTimeUtils.monthFromDateValue(dateValue);
+            dateValue = DateTimeUtils.dateValue(year, month, 1);
+            timeNanos = 0l;
+            break;
+
+        case QUARTER:
+
+            long yearForQuarter = DateTimeUtils.yearFromDateValue(dateValue);
+            int monthForQuarter = DateTimeUtils.monthFromDateValue(dateValue);
+
+            // Round the month to lower quarter
+            if (monthForQuarter >= 10) {
+                monthForQuarter = 10;
+            } else if (monthForQuarter >= 7) {
+                monthForQuarter = 7;
+            } else if (monthForQuarter >= 4) {
+                monthForQuarter = 4;
+            } else {
+                monthForQuarter = 1;
+            }
+
+            dateValue = DateTimeUtils.dateValue(yearForQuarter, monthForQuarter, 1);
+            timeNanos = 0l;
+            break;
+
+        case YEAR:
+
+            long yearForYear = DateTimeUtils.yearFromDateValue(dateValue);
+            dateValue = DateTimeUtils.dateValue(yearForYear, 1, 1);
+            timeNanos = 0l;
+            break;
+
+        case DECADE:
+
+            long yearForDecade = DateTimeUtils.yearFromDateValue(dateValue);
+            yearForDecade = (long) (Math.floor(yearForDecade / 10) * 10);
```

Comment:
Why `Math.floor(double)` is used here? You only converting long to double, invoking the slow function than does nothing with this double because it's already don't have fractional part and converting this double back to long.

The same problem for `CENTURY` and `MILLENNIUM`.

Reply:
Oups, I thought Java rounds up the division result, e.g. `int a = 1/ 3 -> 1`, but in fact no it rounds down `int a = 1 /3 -> 0`;